File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/BlogServiceImpl.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#imports
import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import com.lightbend.lagom.scaladsl.api.Service
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRef
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry

//#imports

trait BlogServiceImpl {
  trait BlogService extends Service {
    def addPost(id: String): ServiceCall[AddPost, String]

    override def descriptor = ???
  }

  //#service-impl
  class BlogServiceImpl(persistentEntities: PersistentEntityRegistry)(implicit ec: ExecutionContext)
      extends BlogService {

    persistentEntities.register(new Post)

    override def addPost(id: String): ServiceCall[AddPost, String] =
      ServiceCall { request: AddPost =>
        val ref: PersistentEntityRef[BlogCommand] =
          persistentEntities.refFor[Post](id)
        val reply: Future[AddPostDone] = ref.ask(request)
        reply.map(ack => "OK")
      }
  }
  //#service-impl

  class BlogServiceImpl2(persistentEntities: PersistentEntityRegistry)(implicit ec: ExecutionContext)
      extends BlogService {

    persistentEntities.register(new Post)

    //#service-impl2
    override def addPost(id: String) = ServiceCall { request =>
      val ref = persistentEntities.refFor[Post](id)
      ref.ask(request).map(ack => "OK")
    }
    //#service-impl2
  }

}

File: docs/manual/java/guide/cluster/code/docs/home/persistence/BlogServiceImpl.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

// #service-impl
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRef;

import javax.inject.Inject;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;
import com.lightbend.lagom.javadsl.api.*;

public class BlogServiceImpl implements BlogService {

  private final PersistentEntityRegistry persistentEntities;

  @Inject
  public BlogServiceImpl(PersistentEntityRegistry persistentEntities) {
    this.persistentEntities = persistentEntities;

    persistentEntities.register(Post.class);
  }

  @Override
  public ServiceCall<BlogCommand.AddPost, String> addPost(String id) {
    return request -> {
      PersistentEntityRef<BlogCommand> ref = persistentEntities.refFor(Post.class, id);
      return ref.ask(request).thenApply(ack -> "OK");
    };
  }
}
// #service-impl
